cbor/shared/src/main/scala/fs2/data/cbor/high/internal/ValueParser.scala

```diff
@@ -0,0 +1,321 @@
+      Pull.output(Chunk.seq(chunkAcc.reverse)) >> rest.pull.uncons.flatMap {
+        case Some((hd, tl)) => parseTextStrings(hd, 0, tl, acc, Nil)
+        case None           => Pull.raiseError(new CborException("unexpected end of input"))
+      }
+    } else {
+      chunk(idx) match {
+        case CborItem.Break =>
+          Pull.pure((chunk, idx + 1, rest, chunkAcc, CborValue.TextString(acc.result())))
+        case CborItem.TextString(text) =>
+          if (acc.size + text.size < 0) {
+            raise(new CborException(s"text string size is limited to max int (${Int.MaxValue}) characters"), chunkAcc)
+          } else {
+            parseTextStrings(chunk, idx + 1, rest, acc.append(text), chunkAcc)
+          }
+        case item =>
+          raise(new CborException(s"unexpected item $item"), chunkAcc)
+      }
+    }
+
+  private def parseTags[F[_]](chunk: Chunk[CborItem],
+                              idx: Int,
+                              rest: Stream[F, CborItem],
+                              tags: CborValue => CborValue,
+                              chunkAcc: List[CborValue])(implicit
+      F: RaiseThrowable[F]): Pull[F, CborValue, Result[F, CborValue => CborValue]] =
+    if (idx >= chunk.size) {
+      Pull.output(Chunk.seq(chunkAcc.reverse)) >> rest.pull.uncons.flatMap {
+        case Some((hd, tl)) => parseTags(hd, 0, tl, tags, Nil)
+        case None           => Pull.raiseError(new CborException("unexpected end of input"))
+      }
+    } else {
+      chunk(idx) match {
+        case CborItem.Tag(tag) => parseTags(chunk, idx + 1, rest, v => tags(CborValue.Tagged(tag, v)), chunkAcc)
+        case _                 => Pull.pure((chunk, idx, rest, chunkAcc, tags))
+      }
+    }
+
+  private val minusOne = BigInt(-1)
+
+  private def parseValue[F[_]](chunk: Chunk[CborItem], idx: Int, rest: Stream[F, CborItem], chunkAcc: List[CborValue])(
+      implicit F: RaiseThrowable[F]): Pull[F, CborValue, Result[F, CborValue]] =
+    parseTags(chunk, idx, rest, identity, chunkAcc).flatMap { case (chunk, idx, rest, chunkAcc, tags) =>
+      if (idx >= chunk.size) {
+        Pull.output(Chunk.seq(chunkAcc.reverse)) >> rest.pull.uncons.flatMap {
+          case Some((hd, tl)) => parseValue(hd, 0, tl, Nil)
+          case None           => Pull.raiseError(new CborException("unexpected end of input"))
+        }
+      } else {
+        chunk(idx) match {
+          case CborItem.False     => Pull.pure((chunk, idx + 1, rest, chunkAcc, tags(CborValue.False)))
+          case CborItem.True      => Pull.pure((chunk, idx + 1, rest, chunkAcc, tags(CborValue.True)))
+          case CborItem.Null      => Pull.pure((chunk, idx + 1, rest, chunkAcc, tags(CborValue.Null)))
+          case CborItem.Undefined => Pull.pure((chunk, idx + 1, rest, chunkAcc, tags(CborValue.Undefined)))
+          case CborItem.SimpleValue(value) =>
+            Pull.pure((chunk, idx + 1, rest, chunkAcc, tags(CborValue.SimpleValue(value))))
```

Comment:
If you bind the pattern `CborItem.SimpleValue(value)` to a variable, you can re-use it instead of allocating a new one here (same in some other cases). Not big, but very easy to do and on the hot path.

Reply:
Not really, it's not the same type, as this converts from items to values :)

Reply:
Details^^
Yeah, I see it now as well!